File: data-plane/files/upload-file.js
const common = require('../../common/common.js')
const util = require('util')
const request = require('request')
const uuidv1 = require('uuid/v1')
const fs = require('fs')

let resourceName = ''  // IoT Hub name
let resourceKey = '' // IoT Hub primary key
let deviceId = 'device11' // IoT Hub device id
let resourceURI = resourceName + '.azure-devices.net'
let tokenExpirationPeriod = 60
let policyKeyName = 'iothubowner'
let apiVersion = '2016-02-03'

let iotHubSasToken = common.getIoTHubSasToken(resourceURI, resourceKey, policyKeyName, tokenExpirationPeriod)
let fileUploadRequestURI = util.format('https://%s/devices/%s/files?api-version=%s', resourceURI, deviceId, apiVersion).toString()
let fileUploadURITemplate = 'https://%s/%s/%s%s'
let notificationURI = util.format('https://%s/devices/%s/files/notifications?api-version=%s', resourceURI, deviceId, apiVersion).toString()

let uploadFile = function (options) {
    // 1. GET FILE STORAGE URI
    request.post(
        {
            url: fileUploadRequestURI,
            body: JSON.stringify({ 'blobName': options.fileName }),
            headers: {
                'Authorization': iotHubSasToken,
                'Content-Type': 'application/json'
            }
        },
        function (error, response, body) {

            console.log(fileUploadRequestURI)
            console.log(response.statusCode)
            console.log(body)

            if (!error && response.statusCode == 200) {

                var fileParts = JSON.parse(body)

                let fileStorageURI = util.format(fileUploadURITemplate,
                    fileParts.hostName,
                    fileParts.containerName,
                    fileParts.blobName,
                    fileParts.sasToken)

                console.log(fileStorageURI)

                // 2. UPLOAD FILE TO BLOB STORAGE

                request.put(
                    {
                        url: fileStorageURI,
                        body: options.body,
                        headers: {
                            'Content-Type': options.contentType,
                            'Content-Length': options.contentLength,
                            'x-ms-blob-type': 'BlockBlob',
                        }
                    },
                    function (error, response, body) {

                        console.log(fileStorageURI)
                        console.log(response.statusCode)
                        console.log(body)

                        if (!error && response.statusCode == 201) {

                            // 3. GET UPLOAD FILE NOTIFICATION
                            request.post(
                                {
                                    url: notificationURI,
                                    body: JSON.stringify({ 'correlationId': fileParts.correlationId }),
                                    headers: {
                                        'Authorization': iotHubSasToken,
                                        'Content-Type': 'application/json'
                                    }
                                },
                                function (error, response, body) {

                                    console.log(notificationURI)
                                    console.log(response.statusCode)
                                    console.log(body)

                                    if (!error && response.statusCode == 204) {
                                        console.log('File successfully uploaded')
                                    } else {
                                        console.log(error)
                                        console.log(response)
                                    }
                                }
                            )
                        } else {
                            console.log(error)
                            console.log(response)
                        }
                    }
                )
            } else {
                console.log(error)
                console.log(response)
            }
        }
    )


}

let bodyText = JSON.stringify({ "device": "device1", "temp": 42 })
uploadFile({
    body: bodyText,
    fileName: uuidv1() + '.json',
    contentType: 'application/json', contentLength: bodyText.length
})

let binaryFile = "iot.png"
uploadFile({
    body: fs.createReadStream(binaryFile),
    fileName: uuidv1() + '.png',
    contentType: 'image/png', contentLength: fs.statSync(binaryFile).size
})

File: data-plane/files/upload-file.py
# pip install requests
import requests
import uuid
from base64 import b64encode, b64decode
from hashlib import sha256
from time import time
from urllib import quote_plus, urlencode
from hmac import HMAC

resourceName = ''  # IoT Hub name
resourceKey = '' # IoT Hub primary key
deviceId = 'device1' # IoT Hub device id
resourceURI = resourceName + '.azure-devices.net'
tokenExpirationPeriod = 60
policyKeyName = 'iothubowner'
apiVersion = '2016-02-03'

def get_iot_hub_sas_token(uri, key, policy_name, expiry=3600):
    ttl = time() + expiry
    sign_key = "%s\n%d" % ((quote_plus(uri)), int(ttl))
    signature = b64encode(HMAC(b64decode(key), sign_key, sha256).digest())

    rawtoken = {
        'sr' :  uri,
        'sig': signature,
        'se' : str(int(ttl))
    }

    if policy_name is not None:
        rawtoken['skn'] = policy_name

    return 'SharedAccessSignature ' + urlencode(rawtoken)

iotHubSasToken = get_iot_hub_sas_token(resourceURI, resourceKey, policyKeyName, tokenExpirationPeriod)
fileUploadRequestURI = 'https://%s/devices/%s/files?api-version=%s' % (resourceURI, deviceId, apiVersion)
fileUploadURITemplate = 'https://%s/%s/%s%s'
notificationURI = 'https://%s/devices/%s/files/notifications?api-version=%s' % (resourceURI, deviceId, apiVersion)

def upload_file(body, fileName, contentType, contentLength):

    # 1. GET FILE STORAGE URI
    fileUploadPartsResponse = requests.post(fileUploadRequestURI, 
        headers={
            'Authorization': iotHubSasToken,
            'Content-Type': 'application/json'
        },
        data = '{ "blobName": "%s"}' % (fileName)
    )

    print(fileUploadRequestURI)
    print(fileUploadPartsResponse.status_code)
    print(fileUploadPartsResponse.text)

    if fileUploadPartsResponse.status_code == 200:
 
        fileUploadParts = fileUploadPartsResponse.json()
        fileUploadURI = fileUploadURITemplate % (fileUploadParts["hostName"], fileUploadParts["containerName"], fileUploadParts["blobName"], fileUploadParts["sasToken"])
        
        # 2. UPLOAD FILE TO BLOB STORAGE
        uploadResponse = requests.put(fileUploadURI, 
            headers={
                'Content-Type': contentType,
                'Content-Length': contentLength,
                'x-ms-blob-type': 'BlockBlob',
                
            },
            data = body
        )

        print(fileUploadURI)
        print(uploadResponse.status_code)
        print(uploadResponse.text)
        
        if uploadResponse.status_code == 201:

            # 3. GET UPLOAD FILE NOTIFICATION
            notificationResponse = requests.post(notificationURI, 
                headers={
                    'Authorization': iotHubSasToken,
                    'Content-Type': 'application/json'
                },
                data = '{"correlationId": "%s" }' % (fileUploadParts["correlationId"])
            )
    
            print(notificationURI)
            print(notificationResponse.status_code)
            print(notificationResponse.text)

bodyText = '{ "device": "device1", "temp": 42 }'
upload_file(bodyText, str(uuid.uuid4()) + ".json", "application/json", str(len(bodyText)))

bodyBinary = open("iot.png", 'rb').read()
upload_file(bodyBinary, str(uuid.uuid4()) + ".png", "image/png", str(len(bodyBinary)))
